src/lib.rs

```diff
@@ -44,19 +49,27 @@ pub fn run_and_output(config: Config) {
 const MAX_FIX_ITERATIONS: usize = 10;
 
 pub fn run(config: Config) -> Vec<ViolationWithContext> {
-    let resolved_rules = config.get_resolved_rules();
-    let aggregated_queries = AggregatedQueries::new(&resolved_rules, &config);
+    let instantiated_rules = config.get_instantiated_rules();
+    let language = SupportedLanguage::Rust;
+    let aggregated_queries = AggregatedQueries::new(&instantiated_rules, language);
     let tree_sitter_grep_args = get_tree_sitter_grep_args(&aggregated_queries);
     let all_violations: Mutex<HashMap<PathBuf, Vec<ViolationWithContext>>> = Default::default();
     let files_with_fixes: AllPendingFixes = Default::default();
+    let instantiated_per_file_rules: AllInstantiatedPerFileRules = Default::default();
     tree_sitter_grep::run_with_callback(
         tree_sitter_grep_args.clone(),
         |capture_info, file_contents, path| {
-            let (rule, rule_listener) =
-                aggregated_queries.get_rule_and_listener(capture_info.pattern_index);
+            let (instantiated_rule, rule_listener_index) =
+                aggregated_queries.get_rule_and_listener_index(capture_info.pattern_index);
             let mut query_match_context =
-                QueryMatchContext::new(path, file_contents, rule, &config);
-            (rule_listener.on_query_match)(capture_info.node, &mut query_match_context);
+                QueryMatchContext::new(path, file_contents, instantiated_rule, &config, language);
+            instantiated_per_file_rules
+                .get(path, instantiated_rule)
+                .on_query_match(
```

Comment:
...and then invoke its listener callback

The way that I came up with to accommodate the fact that there are n listeners per rule (aka n listener callbacks) but they'll need to have visibility of the "per-file run" rule state (so presumably can't just be fed eg a "type-erased" `Arc<dyn RuleInstancePerFile>`) was to make there just be a _single_ `.on_query_match()` trait method (on `RuleInstancePerFile`) that then gets passed the rule listener index so that it can "multiplex" it internally

This would be kind of clunky if rules were manually implementing these traits (you're defining the set of listener queries in a different place than the "multiplex"'d listener callbacks)

But the idea is that we will give them nice proc macros to define it in a way that feels intuitive/"ESLint-like" (that then under the hood implement the multiple traits etc)

At the moment this PR is swinging back away from (plugin-author) ergonomics but hopefully this set of rule traits is a reasonable starting point for being able to accommodate the actual desired behavior of plugins and then we'll take a crack at writing some proc macros to try and also make it ergonomic